### DISCUSSION TITLE
[Feature]: convenient post language buttons

### DISCUSSION THREAD

**Discussion by Some1withgithub** [mastodon/mastodon-ios#1358] [Ideas]
### Description

As a multilingual person, that regularly writes posts in different languages i would like to always post my messages with the correct language setting without needing to open the language setting each time i want to send my post.

### Suggested Solution

Multiple buttons instead of only one to finish(/post) a post with fitting language settings.

Use current arrow button, but have multiple in the top right corner horizontally besides each other with e.g. ISO 639 language codes maybe somewhere underneath the arrow button.
If there is still a version out there that uses the Publish button seen on product screenshot under https://joinmastodon.org/branding, the language code could be attached to it.
Here how it could possible look like: Green is probably "easily" added and orange is optional/for a longer later update.
![suggestion](https://github.com/mastodon/mastodon-ios/assets/161132110/5d63bb30-a313-4988-9bd4-facb73bce251)

(Green arrows)So the normal one stays nearly the same, but with added language code underneath it to indicate the standard setting.
If system language is different compared to standard, add/show second button with differing system language.

(Green arrows)Another extension could be an option setting where frequently used languages can be defined and be represented as the other two (on the screenshot you can see 3 languages English, Japanese and Finnish)

If space is of concern, maybe use another one with a star as a drop down option, that instead of instantly posting lets you choose between your "favorite" languages which you might have defined in options and you use regular. This could confine it to a maximum of 4 buttons: standard/system(or 2 most needed)/favorites/drop down of all.

(Orange)Another optional button could be very similar to the globe symbol seen on the button, but when you confirm, you will post it with the selected language being set(orange arrow in the picture indicates this option).

Number of available buttons could be restricted depending knowledge of UI/UX designer at Mastodon, but i would highly appreciate it, if it would be at least 2 buttons, as it gets annoying to do it manually with the need of multiple actions instead of one.

### Alternatives

Auto detect, but might need more resources and seems overkill.
Or:
Doing nothing and having it as it works for now, resulting in texts not being able to be translated correctly with the deepl integration or users getting annoyed by needing to use multiple actions to set it correctly every post.

### Additional Context

On the bottom of a new reply "window" i have the globe symbol that i have to click and that let's me set the language of my post manually by 1. clicking on it and 2. choosing the language and 3. confirm or decline my choice. 4. i click the button on the top right to publish my post.

That's tedious and i regular forget either what's my standard or to switch to the correct language. As we are in the AI-Hype and i love the translation feature of deepl implemented with mastodon, i could suggest a language auto-detect with it too, but i think my much simpler, less demanding option is better. (not everything has to be auto detected, as i'm fairly capable of doing it myself without really using more resources)

I already see two language options on top of the globe language selection option: Standard and System, which would fit for my demands, but instead of having them only there and needing to go into it each time, i suggest, to add multiple (maybe number restricted to only 1-4) post buttons (arrow on top right corner at least on android) with integrated language options(see suggested solution).

**Comment by zeitschlag** [comment]
Hello @Some1withgithub,

thank you for your suggestion. As I'm confused by your Android-screenshot, let me ask: Do you use Mastodon on iOS or on Android?

You do know that the iOS-app has the option the language a post is written in, right? And that language-thing even check if what you are writing seems to be right correct language and if not, there's a subtle hint on that very language picker as you can see in this screenshot:

![Simulator Screenshot - iPhone 15 Pro - 2024-02-28 at 13 49 28](https://github.com/mastodon/mastodon-ios/assets/2580019/e57e576a-d86f-4811-a5ad-f8d732725152)

Usually I'm posting in German (DE), but as the app thinks this could be wrong, there's the blue dot as a visual indicator. And if you tap on that very language button, you can change the status language and even get a suggestion:

![Simulator Screenshot - iPhone 15 Pro - 2024-02-28 at 13 50 48](https://github.com/mastodon/mastodon-ios/assets/2580019/36def828-5236-44cb-951d-0f8a6524ec45)

So, on iOS we already have that language-auto-detection and I'd suggest to do nothing as using multiple "Publish"-buttons doesn't look like a good option. In case you're using Android, say something and we can transfer this issue over to the correct repository.

**Comment by Some1withgithub** [comment]
Hello @zeitschlag,
thanks for your reply.

I posted here because of: ["Furthermore, we work off of an internal roadmap and aim for feature-parity and consistency with our iOS app. The iOS app is designated as the "primary" between the two, therefore, if you want to request features, please do so in the Mastodon for iOS repository, as you are requesting a feature to be both in iOS and Android (exceptions being system integrations specific to Android)."](https://github.com/mastodon/mastodon-android?tab=readme-ov-file#contributing)

I used the playstore image, as the branding image from joinmastodon uses the version with a button labeled "publish" instead of an arrow, like yours. So maybe it's only iOS version?
I only use android so, if this is an exception to the aforementioned quote, it's probably because the arrow button is not consistent with androids version. I'm fine wherever the feature request will be posted but i could imagine it being here for a consistent UI on iOS and Android.

> You do know that the iOS-app has the option the language a post is written in, right?

Yes so has android.
In the settings i can set a standard language and it recognizes my systems language too.
The globe symbol in the screenshot is basically that, ...

> And that language-thing even check if what you are writing seems to be right correct language and if not, there's a subtle hint on that very language picker as you can see in this screenshot:

...but compared to iOS it does not auto-detect language. So no indicator that it could be the wrong language. Indication would help with forgetting to set the correct language before posting, but could interrupt by calling once attention.

Also, auto-detect still seems overkill and might be faulty, especially in cases where languages get mixed and as you described it, i would still have to use more clicks/actions, as it isn't set automatically when posting?
I prefer visually "worse" things like having two buttons, instead of a clean single button, over aggravating, time consuming need for multiple actions.

I might not have mentioned it directly, but it could be completely optional and in the settings with the standard language, where the standard would be only one button that assumes set standard language but with options to add more buttons mentioned by me.

A thought that came to my mind, just now, about how it could be done/look with "the publish button": Have publish written over buttons labeled with the according language code (i know it's not so pretty as it disrupts the one line) or, if space allows, have publish written before and than those buttons like "Publish: (EN) (DE) (JA)"

Another compromise is to add this functionality with directly posting with another language than standard could be an icon next to the globe/language button on the bottom. As how it is now, the language/globe button in the bottom is only usable for occasionally setting a different language. Otherwise i would not have put time into mentioning alternatives ^^

**Comment by Some1withgithub** [comment]
I feel like that needs even more attention now that i feel like it switched from at least keeping your language setting from first posting it and then editing to me needing to set it each time i make an edit. I also saw a lot of non english toots where deepl translation wanted to translate from english what clearly was not written in english.
